src/modules/vtol_att_control/vtol_type.cpp

```diff
@@ -110,6 +110,66 @@ bool VtolType::init()
 
 }
 
+
+void VtolType::set_weather_vane_yaw_rate()
+{
+	// TODO: do this only in automatic mode
+
+	// First we set the yaw setpoint to the yaw position in order not to have any other control action influencing yaw
+	// Get euler angles
+	matrix::Dcmf R(matrix::Quatf(_v_att->q));
+	matrix::Dcmf R_sp(matrix::Quatf(_v_att_sp->q_d));
+	matrix::Eulerf euler(R);
+	matrix::Eulerf euler_sp(R_sp);
+
+	// compute the yaw error and than rotate the setpoint in a way that yaw is not compensated for
+	float yaw_error = _wrap_pi(euler_sp(2) - euler(2));
+	matrix::Dcmf R_yaw_correction = matrix::Eulerf(0.0f, 0.0f, -yaw_error);
+	matrix::Dcmf R_sp_new = R_sp * R_yaw_correction;
+
+	// save the new modified desired attitude
+	matrix::Quatf(R_sp_new).copyTo(_v_att_sp->q_d);
+
+	// adapt yaw euler angle in uorb message for backward compatibility
+	_v_att_sp->yaw_body = euler(2);
+
+	// direction of desired body z axis represented in earth frame
+	matrix::Vector3f body_z_sp(R_sp_new(0, 2), R_sp_new(1, 2), R_sp_new(2, 2));
```

Comment:
@ivodre Why don't you get the desired body z axis directly from the desired rotation matrix? That is independent from the yaw error.